src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -56,33 +61,301 @@ pub fn copy_to<G, F>(
+                async move {
+                    match mz_aws_util::s3::list_bucket_path(&client, &bucket, &path_prefix).await {
+                        Ok(Some(files)) => {
+                            let files = files
+                                .iter()
+                                .filter(|key| !s3_key_manager.is_sink_object(key))
+                                .collect::<Vec<_>>();
+                            if !files.is_empty() {
+                                Err(anyhow::anyhow!(
+                                    "S3 bucket path is not empty, contains: {:?}",
+                                    files
+                                ))?;
+                            }
+                        }
+                        Ok(None) => {}
+                        Err(e) => Err(e)?,
+                    };
+
+                    info!(%sink_id, %worker_id, "uploading INCOMPLETE sentinel file");
+                    client
+                        .put_object()
+                        .bucket(bucket)
+                        .key(incomplete_sentinel_key)
+                        .send()
+                        .await?;
+
+                    Ok::<(), anyhow::Error>(())
+                }
+            })
+            .wait_and_assert_finished()
+            .await?;
+            Ok::<(), anyhow::Error>(())
+        };
+
+        if is_leader {
+            let res = leader_work().await.map_err(|e| e.to_string());
+            start_handle.give(&start_cap, res).await;
+        };
+    });
+
+    start_stream
+}
+
+/// Renders the 'completion' operator, which expects a `completion_stream`
+/// that contains the result from the upload operations to return to the
+/// `onetime_callback` (see description of `copy_to`).
+///
+/// On the leader worker, this operator waits to see the empty frontier for
+/// the completion_stream and then does some cleanup work before returning
+/// the callback.
+///
+/// This cleanup work removes the INCOMPLETE sentinel file (see description
+/// of `render_initialization_operator` for more details).
+fn render_completion_operator<G, F>(
+    scope: G,
+    connection_context: ConnectionContext,
+    aws_connection: AwsConnection,
+    connection_id: GlobalId,
+    sink_id: GlobalId,
+    s3_key_manager: S3KeyManager,
+    completion_stream: Stream<G, Result<u64, String>>,
+    onetime_callback: F,
+) where
+    G: Scope<Timestamp = Timestamp>,
+    F: FnOnce(Result<u64, String>) -> () + 'static,
+{
     let worker_id = scope.index();
-    let is_leader = usize::cast_from((sink_id, "leader").hashed()) % scope.peers() == worker_id;
+    let num_workers = scope.peers();
+    let leader_id = usize::cast_from((sink_id, "completion").hashed()) % num_workers;
+    let is_leader = worker_id == leader_id;
+
+    let mut builder = AsyncOperatorBuilder::new("CopyToS3-completion".to_string(), scope.clone());
+
+    let mut completion_input = builder.new_disconnected_input(&completion_stream, Pipeline);
```

Comment:
I think you want an exchange pact here so that you hear the counts from all upload operators and add them up

Reply:
The `one_time_callback` thing confused me at first too, but we actually want to call the callback once per worker and then the compute client itself sums up all the responses across workers:
https://github.com/MaterializeInc/materialize/blob/ecd74f9729c0ae2c1634c8afb8c615aa775753b3/src/compute-client/src/service.rs#L473-L502

So I believe using the `Pipeline` pact means the events are passed between the two operators on the same worker, such that we should only receive 1 'completion' message per worker and can call the `one_time_callback` once we receive it (or after doing the cleanup work).

Perhaps it's worth renaming `one_time_callback` to `worker_callback` or something -- I can also add some better comments to explain this.

Reply:
oh! I see. Yes I had not understood that part. Taking a fresh look with this new information in mind